pyriemann/embedding.py

```diff
@@ -0,0 +1,134 @@
+    of a graph where each SPD matrix is a vertex. 
+
+    Parameters
+    ----------
+    metric : string | dict (default: 'riemann')
+        The type of metric to be used for defining pairwise distance between 
+        covariance matrices. 
+
+    """
+
+    def __init__(self, metric='riemann'):
+        """Init."""
+        self.metric = metric
+
+    def fit(self, X):
+        """Fit.
+
+        Do nothing. For compatibility purpose.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_channels)
+            ndarray of SPD matrices.
+
+        Returns
+        -------
+        self : Embedding instance
+        The Embedding instance.
+        """
+        return self
+
+    def transform(self, X, eps=None, tdiff=0.0):
+        """Calculates the coordinates of the embedded points.
+
+        Parameters
+        ----------
+        X :   ndarray, shape (n_trials, n_channels, n_channels)
+              ndarray of SPD matrices.
+        eps:  float (default: None)
+              the scaling of the Gaussian kernel. If none is given
+              it will use the square of the median of pairwise distances between
+              points (same criterium used in the R package implementation).            
+        tiff: float (default: 0.0)
+              diffusion time to be considered. It allows for multiscale analysis
+              but usually tdiff=0.0 gives enough information.
+
+        Returns
+        -------
+        u : ndarray, shape (n_trials, n_trials)
+            ndarray with the embeddings of the covariance matrices.
+        l : ndarray, shape (n_trials)
+            ndarray with the eigenvalues of the diffusion matrix.            
+        """
+        
+        u,l = get_Embedding(X, self.metric, eps, tdiff)
+        
+        return u,l   
+        
+def make_distanceMatrix(points, metric):
+
+    # make matrix with pairwise distances between points
+    Npoints = points.shape[0]
+    distmatrix = np.zeros((Npoints, Npoints))
+    for ii,pi in enumerate(points):
+        for jj,pj in enumerate(points):
+            distmatrix[ii,jj] = distance(pi, pj, metric=metric)
+            
+    return distmatrix       
+
+def make_kernelMatrix(distmatrix, eps):
+    
+    # make kernel matrix from the distance matrix
+    kernel = np.exp(-distmatrix**2/(4*eps))    
+
+    # renormalize the kernel matrix
+    q = np.dot(kernel, np.ones(len(kernel)))
+    kernel_r = np.divide(kernel, np.outer(q,q)) 
+        
+    return kernel_r
+
+def make_transitionMatrix(kernel):
+
+    # normalize rows of the kernel so it becomes a prob transition matrix    
+    d = np.sqrt(np.dot(kernel, np.ones(len(kernel))))
+    P = np.divide(kernel, np.outer(d, d))  
+    
+    return P
+
+def get_Embedding(points, metric, eps=None, tdiff=0):
+    
+    # this implementation follows the algorithm in page 34 of 
+    # Stephane's Lafon PhD thesis "Diffusion Maps and Geometric Harmonics"
```

Comment:
Yeah, I agree with you. 

However, please note that this function actually implements the **Laplacian Eigenmaps** algorithm and not the Diffusion Maps (it is stated in the docstring).

The differences between these two algorithms is really just **one** line: Diffusion maps do an extra normalization in the affinity matrix before doing the eigendecomposition, making it more robust to what the authors call "non uniform sampling of the datapoints". This means that the embedding of your points is less dependent of the probability density used to sample your data points and reflects more precisely the actual geometry of your dataset. 

To be honest, I haven't seen much difference in practice between the results with these two methods, so if you think it would be better to keep things simple, we can use directly the implementation from sklearn. I could make things with a wrapper so that it would be more adapted to covariance matrices and the context of Pyriemann.